V compilation fails compiling code that comapares structs instances

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relvant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.2 bd67b64
**OS:**  Ubuntu 20.04 (Linux)

**What did you do?**
I tried executing following code :
```
    struct User{
	name string
	age int
    }
    mut usr1 := User{'sanath',28}
    mut usr2 := User{'sanath',28}

    if usr1==usr2 {
	println('Same User')
    }
    else {
	println('Not same User')
    }
```

**What did you expect to see?**c
'Same User' printed to console
 
**What did you see instead?**

v run mytry.v
Compilation with tcc failed. Retrying with cc ...

/tmp/v/mytry.1684892087105277150.tmp.c: In function ‘main__main’:
/tmp/v/mytry.1684892087105277150.tmp.c:8340:11: error: invalid operands to binary == (have ‘main__User’ {aka ‘struct main__User’} and ‘main__User’ {aka ‘struct main__User’})
 8340 |  if (usr1 == usr2) {
      |           ^~
...
(Use `v -cg` to print the entire error message)

builder error: 
C error. This should never happen.

If you were not working with C interop, please raise an issue on GitHub:

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

--- a/vlib/v/gen/auto_eq_methods.v
+++ b/vlib/v/gen/auto_eq_methods.v
@@ -18,28 +18,21 @@ fn (mut g Gen) gen_struct_equality_fn(left table.Type) string {
 	fn_builder.writeln('static bool ${ptr_typ}_struct_eq($ptr_typ a, $ptr_typ b) {')
 	for field in info.fields {
 		sym := g.table.get_type_symbol(field.typ)
-		match sym.kind {
-			.string {
-				fn_builder.writeln('\tif (string_ne(a.$field.name, b.$field.name)) {')
-			}
-			.struct_ {
-				eq_fn := g.gen_struct_equality_fn(field.typ)
-				fn_builder.writeln('\tif (!${eq_fn}_struct_eq(a.$field.name, b.$field.name)) {')
-			}
-			.array {
-				eq_fn := g.gen_array_equality_fn(field.typ)
-				fn_builder.writeln('\tif (!${eq_fn}_arr_eq(a.$field.name, b.$field.name)) {')
-			}
-			.map {
-				eq_fn := g.gen_map_equality_fn(field.typ)
-				fn_builder.writeln('\tif (!${eq_fn}_map_eq(a.$field.name, b.$field.name)) {')
-			}
-			.function {
-				fn_builder.writeln('\tif (*((voidptr*)(a.$field.name)) != *((voidptr*)(b.$field.name))) {')
-			}
-			else {
-				fn_builder.writeln('\tif (a.$field.name != b.$field.name) {')
-			}
+		if sym.kind == .string {
+			fn_builder.writeln('\tif (string_ne(a.$field.name, b.$field.name)) {')
+		} else if sym.kind == .struct_ && field.typ.nr_muls() == 0 {
+			eq_fn := g.gen_struct_equality_fn(field.typ)
+			fn_builder.writeln('\tif (!${eq_fn}_struct_eq(a.$field.name, b.$field.name)) {')
+		} else if sym.kind == .array && field.typ.nr_muls() == 0 {
+			eq_fn := g.gen_array_equality_fn(field.typ)
+			fn_builder.writeln('\tif (!${eq_fn}_arr_eq(a.$field.name, b.$field.name)) {')
+		} else if sym.kind == .map && field.typ.nr_muls() == 0 {
+			eq_fn := g.gen_map_equality_fn(field.typ)
+			fn_builder.writeln('\tif (!${eq_fn}_map_eq(a.$field.name, b.$field.name)) {')
+		} else if sym.kind == .function {
+			fn_builder.writeln('\tif (*((voidptr*)(a.$field.name)) != *((voidptr*)(b.$field.name))) {')
+		} else {
+			fn_builder.writeln('\tif (a.$field.name != b.$field.name) {')
 		}
 		fn_builder.writeln('\t\treturn false;')
 		fn_builder.writeln('\t}')

--- a/vlib/v/gen/cgen.v
+++ b/vlib/v/gen/cgen.v
@@ -3036,6 +3036,23 @@ fn (mut g Gen) infix_expr(node ast.InfixExpr) {
 		}
 		g.expr(node.right)
 		g.write(')')
+	} else if node.op in [.eq, .ne] && left_sym.kind == .struct_ && right_sym.kind == .struct_ {
+		ptr_typ := g.gen_struct_equality_fn(left_type)
+		if node.op == .eq {
+			g.write('${ptr_typ}_struct_eq(')
+		} else if node.op == .ne {
+			g.write('!${ptr_typ}_struct_eq(')
+		}
+		if node.left_type.is_ptr() {
+			g.write('*')
+		}
+		g.expr(node.left)
+		g.write(', ')
+		if node.right_type.is_ptr() {
+			g.write('*')
+		}
+		g.expr(node.right)
+		g.write(')')
 	} else if node.op in [.key_in, .not_in] {
 		if node.op == .not_in {
 			g.write('!')

--- a/vlib/v/tests/struct_equality_test.v
+++ b/vlib/v/tests/struct_equality_test.v
@@ -1,0 +1,15 @@
+struct User {
+	name string
+	age  int
+}
+
+fn test_struct_equality() {
+	mut usr1 := User{'sanath', 28}
+	mut usr2 := User{'sanath', 28}
+	if usr1 == usr2 {
+		println('Same User')
+	} else {
+		println('Not same User')
+	}
+	assert usr1 == usr2
+}

